packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -184,20 +186,40 @@ export class LoaderContainerTracker implements IOpProcessingController {
                         // Only write it out once
                         waitingSequenceNumberSynchronized = true;
                         debugWait("Waiting for sequence number synchronized");
-                        await this.waitForAnyInboundOps(containersToApply);
+                        await timeoutAwait(this.waitForAnyInboundOps(containersToApply), {
+                            durationMs: timeoutDuration - (Date.now() - start),
+                            errorMsg: "Timeout on waiting for sequence number synchronized",
+                        });
                     }
                 } else {
                     waitingSequenceNumberSynchronized = false;
-                    await this.waitForPendingClients(pendingClients);
+                    await timeoutAwait(this.waitForPendingClients(pendingClients), {
+                        durationMs: timeoutDuration - (Date.now() - start),
+                        errorMsg: "Timeout on waiting for pending join or leave op",
+                    });
                 }
             } else {
                 // Wait for all the containers to be saved
                 // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                 debugWait(`Waiting container to be saved ${dirtyContainers.map((c) => this.containers.get(c)!.index)}`);
+                const remainedDuration = timeoutDuration - (Date.now() - start);
                 waitingSequenceNumberSynchronized = false;
                 await Promise.all(dirtyContainers.map(async (c) => Promise.race(
-                    [new Promise((resolve) => c.once("saved", resolve)),
-                    new Promise((resolve) => c.once("closed", resolve))],
+                    [timeoutPromise(
+                        (resolve) => c.once("saved", () => resolve()),
+                        {
+                            durationMs: remainedDuration / dirtyContainers.length,
+                            errorMsg: "Timeout on waiting a container to be saved",
+                        },
+                    ),
+                    timeoutPromise(
+                        (resolve) => c.once("closed", () => resolve()),
+                        {
+                            durationMs: remainedDuration / dirtyContainers.length,
+                            errorMsg: "Timeout on waiting a container to be closed",
```

Comment:
i don't think we should actually timeout here. this is a fail safe to catch when container close, rather than reach the saved state, and since these are both in a promise.race, the first completing promise will win, so if we only timeout the "saved" case the only time this case will occur is if the close happens before the save or before the save timeout